ibis/pandas/execution.py

```diff
@@ -0,0 +1,438 @@
+
+def reduction_to_lambda(expr):
+    op = expr.op()
+
+    def reduction(
+        df,
+        column=op.args[0]._name,
+        method_name=type(op).__name__.lower(),
+        new_name=expr._name,
+    ):
+        method = getattr(df[column], method_name)
+        return method().rename(new_name)
+    return reduction
+
+
+@execute_node.register(ops.Aggregation, pd.DataFrame)
+def execute_aggregation_dataframe(op, data, scope=None):
+    assert op.agg_exprs
+    assert op.by
+
+    if op.having:
+        raise NotImplementedError('having expressions not yet implemented')
+
+    if op.sort_keys:
+        raise NotImplementedError(
+            'sorting on aggregations not yet implemented'
+        )
+
+    predicates = op.predicates
+    if predicates:
+        data = data.loc[
+            functools.reduce(
+                operator.and_, (execute(p, scope) for p in predicates)
+            )
+        ]
+
+    gb = data.groupby([execute(by, scope) for by in op.by])
+
+    aggs = [reduction_to_lambda(agg_expr) for agg_expr in op.agg_exprs]
+    pieces = [func(gb) for func in aggs]
+    agg_df = pd.concat(pieces, axis=1)
+
+    return agg_df.reset_index()
+
+
+@execute_node.register(ops.Reduction, pd.Series, (pd.Series, type(None)))
+def execute_reduction_series_mask(op, data, mask, scope=None):
+    operand = data[mask] if mask is not None else data
+    return getattr(operand, type(op).__name__.lower())()
+
+
+@execute_node.register(ops.StandardDev, pd.Series, (pd.Series, type(None)))
+def execute_standard_dev_series(op, data, mask, scope=None):
+    return (data[mask] if mask is not None else data).std()
+
+
+@execute_node.register(ops.Variance, pd.Series, (pd.Series, type(None)))
+def execute_variance_series(op, data, mask, scope=None):
+    return (data[mask] if mask is not None else data).var()
+
+
+@execute_node.register((ops.Any, ops.All), pd.Series)
+def execute_any_all_series(op, data, scope=None):
+    return getattr(data, type(op).__name__.lower())()
+
+
+@execute_node.register(ops.Not, (bool, np.bool_))
+def execute_not_bool(op, data, scope=None):
+    return not data
+
+
+_JOIN_TYPES = {
+    ops.LeftJoin: 'left',
+    ops.InnerJoin: 'inner',
+    ops.OuterJoin: 'outer',
+}
+
+
+@execute_node.register(ops.Join, pd.DataFrame, pd.DataFrame)
+def execute_join(op, left, right, scope=None):
+    try:
+        how = _JOIN_TYPES[type(op)]
+    except KeyError:
+        raise NotImplementedError('{} not supported'.format(type(op).__name__))
+
+    left_on = []
+    right_on = []
+
+    for predicate in op.predicates:
+        pred_op = predicate.op()
+        if not isinstance(pred_op, ops.Equals):
+            raise TypeError(
+                'Only equality join predicates supported with pandas'
+            )
+        left_key = pred_op.left._name
+        right_key = pred_op.right._name
+        left_on.append(left_key)
+        right_on.append(right_key)
+
+    return pd.merge(left, right, how=how, left_on=left_on, right_on=right_on)
```

Comment:
could inconsistencies in column names cause any issues? I can't think of any off top of head